middleware/bos-authentication.js

```diff
@@ -0,0 +1,326 @@
+var _ = require('lodash');
+var base64URL = require('base64url');
+
+var log;
+var loader;
+var httpMethods = ['get', 'put', 'post', 'delete', 'options', 'head', 'patch'];
+//map of middleware ids to 'true' meaning that they have been initialized
+var middlewareInitMap = {};
+
+module.exports = {
+    init : init
+};
+
+function init(app, logger, serviceLoader, swagger) {
+    log = logger;
+    loader = serviceLoader;
+    _.forEach(swagger.getSimpleSpecs(), function (api, name) {
+        var basePath = api.basePath || '';
+        /* apply security requirements to each route path*/
+        _.keys(api.paths).forEach(function (path) {
+            var pathObj = api.paths[path];
+            var routePath = basePath + _convertPathToExpress(path);
+
+            //loop for http method keys, like get an post
+            _.keys(pathObj).forEach(function (method) {
+                if (_.contains(httpMethods, method)) {
+                    var operation = pathObj[method];
+                    if (operation['security']) {
+                        operation['security'].forEach(function (securityReq) {
+                            _.forOwn(securityReq, function (scopes, securityDefn) {
+                                _applySecurityRequirement(app, method, routePath, securityDefn,
+                                    api.securityDefinitions[securityDefn],
+                                    /*operation['x-bos-permissions'][securityReq],*/
+                                    scopes);
+                            });
+                        });
+                    }
+                    if (operation['x-bos-security']) {
+                        operation['x-bos-security'].forEach(function (securityReq) {
+                            _.forOwn(securityReq, function (scopes, securityDefn) {
+                                _applyCustomSecurityRequirement(app, method, routePath, securityDefn,
+                                    api['x-bos-securityDefinitions'][securityDefn],
+                                    /*operation['x-bos-permissions'][securityReq],*/
+                                    scopes);
+                            });
+                        });
+                    }
+                }
+            });
+        });
+    });
+}
+
+function _applyCustomSecurityRequirement(app, method, route, securityReq,
+                                   securityDefn, /*requiredPermissions,*/ requiredScopes) {
+    //load security def middleware
+    if (securityDefn['x-bos-middleware']) {
+        var customAuthMiddleware = loader.getConsumer('middleware', securityDefn['x-bos-middleware']);
+        if (!customAuthMiddleware) {
+            loader.loadConsumerModules('middleware',
```

Comment:
it won't do a reload from disk since it is using node's require cache behind the scenes, but still the code is more efficient if you don't attempt to reload the middleware.